extensions/vscode/lsp-client/src/extension.ts

```diff
@@ -51,12 +46,316 @@ export function activate(context: ExtensionContext) {
+
+	private readonly _panel: vscode.WebviewPanel;
+	private _disposables: vscode.Disposable[] = [];
+
+	public static createOrShow() {
+		const column = vscode.window.activeTextEditor
+			? vscode.window.activeTextEditor.viewColumn
+			: undefined;
+
+		// If we already have a panel, show it.
+		if (TimelinePanel.currentPanel) {
+			TimelinePanel.currentPanel._panel.reveal(column);
+			return;
+		}
+
+		// Otherwise, create a new panel.
+		const panel = vscode.window.createWebviewPanel(
+			TimelinePanel.viewType,
+			'Timeline',
+			column || vscode.ViewColumn.One,
+			getWebviewOptions(),
+		);
+
+		TimelinePanel.currentPanel = new TimelinePanel(panel);
+	}
+
+	public static revive(panel: vscode.WebviewPanel) {
+		TimelinePanel.currentPanel = new TimelinePanel(panel);
+	}
+
+	private constructor(panel: vscode.WebviewPanel) {
+		this._panel = panel;
+
+		// Set the webview's initial html content
+		this._update();
+
+		// Listen for when the panel is disposed
+		// This happens when the user closes the panel or when the panel is closed programmatically
+		this._panel.onDidDispose(() => this.dispose(), null, this._disposables);
+
+		// Update the content based on view changes
+		this._panel.onDidChangeViewState(
+			e => {
+				if (this._panel.visible) {
+					this._update();
+				}
+			},
+			null,
+			this._disposables
+		);
+
+		// Handle messages from the webview
+		this._panel.webview.onDidReceiveMessage(
+			message => {
+				switch (message.command) {
+					case 'alert':
+						vscode.window.showErrorMessage(message.text);
+						return;
+				}
+			},
+			null,
+			this._disposables
+		);
+	}
+
+	private _getDB() {
+		return new sqlite3.Database(__dirname + '/../../../../wake.db', sqlite3.OPEN_READONLY, (err) => {
+			if (err) {
+				return console.error(err.message);
+			}
+		});
+	}
+
+	public refresh() {
+		let db = this._getDB();
+		this._queryDB(db, (jobsJson: JobReflection[], accessesJson: { type: number, job: number }[]) => {
+			let message = {
+				jobReflections: jobsJson,
+				fileAccesses: accessesJson
+			}
+			// Send a message to the webview.
+			this._panel.webview.postMessage(message);
+			db.close();
+		});
+	}
+
+	public dispose() {
+		TimelinePanel.currentPanel = undefined;
+
+		// Clean up our resources
+		this._panel.dispose();
+
+		while (this._disposables.length) {
+			const x = this._disposables.pop();
+			if (x) {
+				x.dispose();
+			}
+		}
+	}
+
+	private _update() {
+		this._panel.title = 'Timeline';
+
+		let db = this._getDB();
+		try {
+			let data = fs.readFileSync(__dirname + '/../../../../share/wake/html/timeline.html', { encoding: 'utf8' }).toString();
+			this._queryDB(db, (jobsJson: JobReflection[], accessesJson: { type: number, job: number }[]) => {
+				this._panel.webview.html = this._formatString(data, JSON.stringify(jobsJson), JSON.stringify(accessesJson));
+				db.close();
+			});
+		} catch (err) { }
+		return;
+	}
+
+	private _formatString(s: string, ...args: string[]): string {
```

Comment:
What is this function doing? I've stared at it for a little bit, but it's pretty difficult to grok, and it seems to be making a lot of strong assumptions about the arguments being passed in.

Once again, I don't know much about VSCode extensions, but is there some other way of passing dynamic data to a web view in VSCode? If so, I think it's worth refactoring the HTML file to be better suited for passing data into it from VSCode.

Reply:
I had the same reaction and requested a comment but it hit me before I could type it out. It formats a string with things like {0} {1} etc... in them

```
_formatString("{0}, {1}!", "Hello", "World")
```

It's replacing `{d+}` with the corresponding things in the args array.

Since both of us hit this we should likely add a comment it.

Reply:
Ah, so the `{` is treated as literally a `{` character, not as the special regexp syntax for repeating a pattern multiple times. Does the `{` not need to be escaped here because it's at the very beginning of the regexp and therefore it is impossible for it to be attached to a pattern?

I think if it's not harmful, I'd suggest both 1) explicitly escaping the `{` so that a reader doesn't mistake it for its special meaning, and 2) documenting it.

But I think ideally, if VSCode has a different way of passing data to a web view besides dynamically generating JSON strings that get embedded into HTML, that would be better than all of this.

Reply:
Removed